src/memory.cpp

```diff
@@ -76,44 +100,50 @@ std::pair<expr, expr> SingleArrayMemory::load(
 }
 
 
-MultipleArrayMemory::MultipleArrayMemory(unsigned int numBlocks):
-    Memory(ulog2(numBlocks), numBlocks) {
-  for (unsigned i = 0; i < numBlocks; ++i) {
+MultipleArrayMemory::MultipleArrayMemory(
+    unsigned int numGlobalBlocks, unsigned int nunLocalBlocks):
+  Memory(numGlobalBlocks, nunLocalBlocks, ulog2(numGlobalBlocks + nunLocalBlocks)) {
+  for (unsigned i = 0; i < getNumBlocks(); ++i) {
     auto suffix = [&](const string &s) {
       return s + to_string(i);
     };
     arrays.push_back(ctx.constant(suffix("array").c_str(),
         ctx.array_sort(Index::sort(), Float::sort())));
     writables.push_back(ctx.bool_const(suffix("writable").c_str()));
     numelems.push_back(ctx.bv_const(suffix("numelems").c_str(), Index::BITS));
+    isGlobals.push_back(ctx.bool_val(i < numGlobalBlocks));
   }
 }
 
 expr MultipleArrayMemory::itebid(
     const expr &bid, function<expr(unsigned)> fn) const {
-  assert(numBlocks > 0);
+  assert(getNumBlocks() > 0);
   assert(bid.get_sort().is_bv() && bid.get_sort().bv_size() == getBIDBits());
 
   uint64_t const_bid;
   if (bid.is_numeral_u64(const_bid))
     return fn(const_bid);
 
   const unsigned bits = bid.get_sort().bv_size();
-  unsigned curbid = numBlocks - 1;
-  expr val = fn(curbid);
 
-  while (curbid) {
-    curbid--;
-    val = z3::ite(bid == ctx.bv_val(curbid, bits), fn(curbid), val);
-  }
+  expr globalExpr = fn(0);
+  for (unsigned i = 1; i < numGlobalBlocks; i ++)
+    globalExpr = z3::ite(bid == ctx.bv_val(i, bits), fn(i), globalExpr);
+
+  if (currLocalBlocks == 0)
+    return globalExpr;
 
-  return val;
+  expr localExpr = fn(numGlobalBlocks);
+  for (unsigned i = numGlobalBlocks + 1; i < getNumBlocks(); i ++)
+    localExpr = z3::ite(bid == ctx.bv_val(i, bits), fn(i), localExpr);
```

Comment:
Hmm.. do we need to separate these two loops? Maybe we keep them as one as before.

Reply:
At first, I think it's helpful when encoding bid which pointing local memblock (because #local memblock is usually smaller than #global memblock). 
But considering `isGlobalBlock(bid)` is also expressed by long z3::ite terms, my thought might not be true.
I'll fix it using single loop.